### ISSUE TITLE
Move the logic from Cylc commands/scripts to the cylc package

### ISSUE THREAD

**Issue by kinow** [cylc/cylc-flow#2802]
**For reference: [python packaging - Command Line Scripts](https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html)**

I have [a branch](https://github.com/cylc/cylc/compare/master...kinow:make-cylc-a-module?expand=1) with a work-in-progress `setup.py` for Cylc (if you have a look at the branch, skip the deleted files and don't mind the number of commits).

In the `setup.py`, we have the new list of dependencies for Cylc, like `isodatetime`, and `jinja2`, matching versions required. And there is also an entry for **`scripts`**. The `scripts` option is configured to list all files under the `bin/` folder.

What it means, is that running `pip install cylc` (not exactly as we haven't published anything there... but just for example) would install the module `cylc` (i.e. contents of `lib/cylc`), its dependencies (e.g. versions needed for `jinja2`, and `isodatetime`), and also put the Cylc scripts in the `$PATH`.

Right now users are asked to do that as a part of the installation process, but with this approach that won't be necessary anymore. The scripts would be installed with the new module in the user environment.

Furthermore, users would be able to create `virtualenv`'s, `conda` environments, or if the deb/rpm packaging works, they could use that option too. And that way they would be able to manage multiple versions of Cylc. 

It is also possible to achieve the same with containers, but containers are at a higher level. The option with `pip` requires simply a Python installation, `pip`, and `virtualenv` to support multiple versions (or in the supercomputer perhaps that `load module` trick...).

One problem with this approach, however, is that it becomes clear that the scripts under the `bin/` folder contain too much logic.

This ticket suggests to move that logic to the Cylc module (i.e. under `lib/cylc`), within appropriate objects for encapsulation. Write tests (which is harder right now), and then later replace the `scripts` by the Python [entry point](https://packaging.python.org/specifications/entry-points/): [`console_scripts`](https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-console-scripts-entry-point).

The `console_scripts` allow to expose **functions** as scripts.

In other words, `cylc scan` would be a [thin shim](https://en.wikipedia.org/wiki/Shim_(computing)), calling actually a function like `SomeCylcObject.scan(*args)`.

The clear downside is the complexity and risk of this change.

But here's a list of benefits:

- We can test the scripts
- IMO, it would be easier to maintain
- Easier to expose, deprecate, replace scripts, as we would change simply the function exposed in the entry point
- Users that did a `import cylc` in a Jupyter Notebook, for example, would be able to write code that calls that same function, without having to fork a process
- The new Web GUI wouldn't have to fork processes (i.e. start a new process, allocate memory, file handlers, I/O, etc), it would simply import files from the syspath, within the Python world, and execute exactly what a user does with `cylc scan` or any other command (Probably not entirely true... we could still re-use the `main() method` in most cases?...)

This can be done after the new Web GUI. I would prefer to try to implement this before, as I think the code would look simpler... but on the other hand, if we are not able to implement it right now, I could benchmark the current approach with the scripts, and compare against this approach and validate the last item.

Cheers
Bruno

**Comment by matthewrmshin** [comment]
Agree that most logic in `bin/` should be moved under `lib/cylc/`.

**Comment by hjoliver** [comment]
Also agree - everything above sounds good to me.  We could do it now, if you think it wouldn't delay us too long?

**Comment by kinow** [comment]
I'm pondering whether I should give it a try while we have some time before the work on Python 3 on isodatetime/Cylc. Will choose a command that I am familiar like `run` or `scan` to see how it goes. Thanks!!

**Comment by kinow** [comment]
Interestingly `memory_profiler` seems to indicate less memory used with the option with `subprocess`.

```
Filename: test.py

Line #    Mem usage    Increment   Line Contents
================================================
     7     28.6 MiB     28.6 MiB   @profile()
     8                             def test():
     9     28.6 MiB      0.0 MiB       proc = Popen(
    10     28.6 MiB      0.0 MiB           cmd,
    11     28.7 MiB      0.0 MiB           stdin=open(os.devnull), stdout=PIPE, stderr=PIPE)
    12     28.7 MiB      0.0 MiB       if proc.wait():
    13                                     print("Error executing command: %s" % proc.communicate()[1])
    14                                 else:
    15     28.7 MiB      0.0 MiB           comm = proc.communicate()
    16     28.7 MiB      0.0 MiB           print("Done! %s\n%s" % (comm[0], comm[1]))


Filename: test2.py

Line #    Mem usage    Increment   Line Contents
================================================
     5     42.7 MiB     42.7 MiB   @profile
     6                             def test2():
     7     42.7 MiB      0.0 MiB       cmd = ['run', 'suite', '--no-detach', '--verbose']
     8                             
     9     42.7 MiB      0.0 MiB       sys.argv = cmd
    10                             
    11                             
    12     53.6 MiB     10.9 MiB       scheduler_cli.main()
```

![figure_1](https://user-images.githubusercontent.com/304786/48596182-bf746580-e9bc-11e8-9e0a-60efb1b0ecf1.png)

![figure_2](https://user-images.githubusercontent.com/304786/48596184-c307ec80-e9bc-11e8-8292-3d2b143c0d78.png)

**Comment by hjoliver** [comment]
> While reviewing #2809 I found that scheduler.py uses subprocess.Popen to execute Cylc commands.

@kinow - this only happens on #2809, no? (in which case, shouldn't this be a comment on that PR?)

**Comment by kinow** [comment]
@hjoliver I think the `scheduler.py` in the `master` branch does that too. It will put commands in the `SubProcPool` to be executed via `subprocess.Popen` to execute `cylc ....`, like sending messages.

I have a piece of paper I was tracking down some calls in a suite at work. I can confirm tomorrow, but I think I had a branch somewhere experimenting with `functools.partials` as a naive/poor-man's replacement for the `subprocess.Popen`, to measure what happens with performance/monitoring/cpu/etc.

That's why I kept this part here. I think the work on #2809 is all good, as far as I could tell. It works like the other parts of the system. I am still planning to test it in my local environment with a Docker container, then send a +1 :star: 

Another by-product of this test, I hope, will be an experiment - or maybe an example - to show how to use Docker with Travis-CI. This way we could have tests with real remote environments if necessary.

**Comment by kinow** [comment]
(but re-reading the issue description here, maybe it would be better to put it in another issue, sorry)

**Comment by hjoliver** [comment]
@kinow - OK understood - I was just taking your statement too literally! (`scheduler.py` itself does not execute cylc commands in subprocesses ... but it certainly gets other modules to do it).

**Comment by matthewrmshin** [comment]
Hi @kinow `cat-log` relies on `KeyboardInterrupt` to terminate sub-processes such as `tail -F ...` (which would run forever until signalled.)

**Comment by kinow** [comment]
But I think it prints the error in the consolr output after you press ctrl+c. Or another one does that. I noticed some commands catch KeyboardInterrupt and exit nicely, while others exit and also let the user see the errors in the terminal

**Comment by matthewrmshin** [comment]
Yes, fully agreed on consistency with `KeyboardInterrupt` behaviour.

Just re-reading some of the previous comments...

On `cylc.scheduler` running sub-processes. It is currently a very cheap way to run things in a pool of parallel *workers*. (Processes in Unix are also more visible and easier to manage via the operating system. E.g. They are more independent, easier to understand and killable by users.) We have tried threading in the past, but we can perhaps try `asyncio` in the future.

The biggest commands are:
* `cylc jobs-submit` - for job submission, which can in theory run as a function in the background if submitting local jobs, but will require running commands on remote hosts via SSH. The current command encapsulates everything in a uniform way. The command does various I/O and calls out to the workload manager submission command e.g. `qsub`. Turning this into an async function is going to be interesting.
* `cylc jobs-poll` - for job poll. Ditto.
* `cylc jobs-kill` - for job kill. Ditto - but should not be used as much.
* Event mail - Using the `mail` command here for uniformity with the other stuff and also because `mail` is easier to configure than Python's `smtplib` module.
* Job logs retrieval - Using the `rsync` command - not aware of a pure Python equivalent.
* Event handler scripts - These are shell scripts at the moment. Until we allow them to be configured as Python logic (see #1962), we are a bit stuck here.
* External triggers - some can probably be implemented in pure Python, e.g. via suite network API for external suite polling.
* Anything else?

**Comment by sgaist** [comment]
Following the discutions on #3393, about going with entry_points and a central command with subcommands, it looks like click could be used for that but that would require a multistep migration.
I'd propose to start by moving the "pure python" scripts from bin to a scripts subfolder (i.e. cycle/flow/scripts) and create the corresponding entry_points

That would keep the current workflow as is in terms of implementation.

In a second phase, move the cylc script back to python using click and then make all the other commands sub commands of the main click application.

**Comment by kinow** [comment]
This sounds like a plan! Are you interested in working on a pull request for this?

For personal projects I use click as I find it easier to use. But it would be good if we could implement this first without throwing away the argpase parts.

Would it be ppssible too @sgaist?

**Comment by hjoliver** [comment]
Click does look nice https://click.palletsprojects.com/en/7.x/

@oliver-sanders - do you have an opinion on this?  (For one thing, I'm not sure if there'd be any conflict with our current use of CLI colors?).

**Comment by oliver-sanders** [comment]
Click does indeed look nice, another really nice one is [fire](https://github.com/google/python-fire).

* Our use of colour is portable.
* The `cli_function` wrapper would need to be re-written for whatever parser we move to but that's not much work (and anticipated anyway).

There was some discussion on arg parsing in #2972, we aught to move away from optparse, I'm happy to move to whatever, but here are a couple of requirements:

* Need to be able to programatically generate argument parsers to api-on-the-fly [1].
* Need to be able to auto-document in Sphinx.
* Not really interested in completely re-writing if only for syntactic sugar.

The easiest conversion would be argparse, not the most exciting but it ticks all the boxes. It would be nice to delete `cylc.flow.option_parsers` and start afresh though.

[1] We want to generate a good chunk of the CLI from GraphQL mutations.

**Comment by hjoliver** [comment]
(interestingly click is based on optparse, not argparse: https://click.palletsprojects.com/en/7.x/why/#why-not-argparse)

**Comment by kinow** [comment]
Today I spent some time working with `conda` and `conda-build`. They have a similar setup to what is proposed here and with #2959 

In conda they have the `conda` command, that handles parsing (with `argparse` FWIW), initializing logging, and calling the subcommand.

Looks like conda-build extends the possible commands from `conda`. If you have downloaded Anaconda, and try `conda build`, it will fail, as there is no subcommand `build`.

Once you install `conda-build`, then `conda build` becomes available. That's the same idea that we could implement in `cylc`. So that `cylc test-battery` doesn't exist when you are running Cylc Flow on your HPC cluster, cloud, personal desktop, etc.

But for a development environment, you will probably want that, and then maybe we can have a `pip install cylc-test-battery` that brings that command :tada:

**Comment by sgaist** [comment]
Moving the python scripts as well as the "bash forwarder" to entry_points is pretty easy however the cylc bash script itself does some "in-folder" checks as well as short name handling that make the functional tests fail.

**Comment by kinow** [comment]
> Moving the python scripts as well as the "bash forwarder" to entry_points is pretty easy however the cylc bash script itself does some "in-folder" checks as well as short name handling that make the functional tests fail.

Hmm, I wonder if that's the reason we haven't touched it yet... So we can either change that shell script, or pythonize the cylc shell script I guess? If the latter is not too hard, that could also help us in migrating everything to python.

**Comment by hjoliver** [comment]
> pythonize the cylc shell script 

That would not be difficult.  It used to be Python anyway, but I recall we convinced ourselves there was a significant performance advantage in using bash for the main `cylc` command.

 @oliver-sanders - does you comment here suggest there was no great advantage and we can just revert to Python? https://github.com/cylc/cylc-flow/issues/3041#issuecomment-536936315

**Comment by oliver-sanders** [comment]
**tldr;** Yeah, just do it.

> That would not be difficult. It used to be Python anyway

Sorry, the conversion was kinda my fault. The old `cylc` Python script had loads of help strings and a fair bit of logic which placed a bit of an overhead on simple commands. In hindsight simply factoring the help out into a separate Python module was probably enough.

A bit of context for those not so familiar with Cylc; some of the Cylc commands are interactive, performance isn't so important here. However, other commands are called non-interactively and, potentially, a large number of times. In these cases keeping overheads to a minimum has its advantages. For example `cylc message` is called multiple times by each job, suites might submit tens of thousands of jobs per day and sites might have hundreds of running suites.

**Comment by matthewrmshin** [comment]
(The performance depends on where Python and Cylc are installed, and the frequency of the hit on those locations and by what user environments. There is also an assumption that calling `/bin/bash` and other GNU coreutils are cheap.)

**Comment by sgaist** [comment]
I have a proof of concept branch that uses the original cylc python script with some fixes to make it run with Python 3.

Some tests failing but I haven't got time yet to fully dig into the internals of cylc to ensure what the errors are about. It might be due also to the fact that I'm running this on macOS.

**Comment by kinow** [comment]
>I have a proof of concept branch that uses the original cylc python script with some fixes to make it run with Python 3.

Great! That was super quick.

>Some tests failing but I haven't got time yet to fully dig into the internals of cylc to ensure what the errors are about.

If you would like, feel free to share your branch link here, and one of us can take a look. If you have unit tests failing, I'd be happy to take a look and try to come up with the fixes. The functional tests may require another dev with more experience with them. But still, that may be easier for one of us to spot how to fix the tests or the code, saving you some time - unless you want to dig into these tests and fix them.

>It might be due also to the fact that I'm running this on macOS.

What you can do, is push to your fork on GitHub, and activate Travis CI, if you haven't already done that. That way you can leave Travis to run all the test battery and report back on test pass/fail as well as coverage, etc.

We recently got another user that [reported trying to use Cylc on macOS](https://cylc.discourse.group/t/cylc-8-0a1-preview-release-via-conda/149/5), though I think s/he just used another computer with Linux in the end. I think we should be able to support macOS too, maybe with some tweaking. But that would be for another issue.

Thanks!
Bruno

ps: if you need to ask any quick questions while working on the code, feel free to stop by our Riot chat too: https://riot.im/app/#/room/#cylc-general:matrix.org

**Comment by hjoliver** [comment]
> (The performance depends on where Python and Cylc are installed, and the frequency of the hit on those locations and by what user environments. There is also an assumption that calling `/bin/bash` and other GNU coreutils are cheap.)

So do you have a strong opinion either way on this @matthewrmshin ?

**Comment by hjoliver** [comment]
@sgaist - if you've gone back to the original cylc python script from several years ago, we'll need to be a) careful about changes made since then in the newer bash script; and b) take a very close look at performance in light of comments above.  That script is pretty ancient, and probably not well coded. (On the other hand, it is fundamentally quite simple, so it should be easy to streamline it).

**Comment by matthewrmshin** [comment]
> So do you have a strong opinion either way on this @matthewrmshin ?

For that statement, I have no opinion either way. Just want to provide extra context.

The reality is that apart from the occasional `cylc help` or `cylc --version` commands, all other `cylc ...` commands will end up calling a Python `cylc-...` command, so there is really not much gain either way. The shell script did have the advantage of being able to set up the environment for Python where necessary. However, now that the project is using package managers, the usefulness of the shell script has totally diminished.

For simplicity (and for the future Python API), I fully support moving towards a pure Python system.

**Comment by oliver-sanders** [comment]
> I think we should be able to support macOS too, maybe with some tweaking

Let us know what the issues are, I've fixed a couple of simple Mac OS bugs in the past:

* The Cylc tests will require coreutils (install via brew) accessible without the `g` prefix (e.g. `ls` rather than `gls` - see the homebrew coreutils docs for more info).
* If you see issues which look like network, domain resolution, etc, this is something Mac OS users have hit before, it's a DNS configuration problem which I don't think we know how to solve yet.

**Comment by oliver-sanders** [comment]
> we'll need to be a) careful about changes made since then in the newer bash script

The old script may be good for a proof of concept but there have been a few changes, also the old Python script wasn't great.

**Comment by hjoliver** [comment]
That's a consensus then :+1: back to Python for `bin/cylc` but maybe start from the current Bash script rather than the (very) old and (not very good) Python script.

**Comment by sgaist** [comment]
As requested, the branch is https://github.com/sgaist/cylc-flow/tree/move_to_entry_points

I have made a minimal cylc script using click since it seems to be one of the possible future framework for building the command set.

Note that it's not optimal currently as for example click provides some tools to handle abbreviated commands.

On a side note, did you consider using [pre-commit](https://pre-commit.com) for checking/fixing code before committing ? There are some nice tools like black, bandit, flake8, etc. that could be run directly on the developer machine and might help keeping the code base tidy without having to redo commits.

**Comment by kinow** [comment]
Might be a good idea the pre commit hooks. JupyterHub and/or pandad I think had some. Feel free to create an issue/PR for it :)

**Comment by kinow** [comment]
(Working on the frontend today, but will take a look at the branch as soon as I'm done over there. Thanks @sgaist!!!)

**Comment by hjoliver** [comment]
@sgaist - good idea on the pre-commit hooks.  We do run bandit, and style checking, etc., via codacy and github, but that errors revealed there do require "redoing commits" as you say. :+1:

**Comment by hjoliver** [comment]
@sgaist - I took a quick look at your branch.  Initial comments:
- nice
- "refactor script and simplify it using click" - wow!
- the result after `pip install` seems fully functional (I ran a suite, which tests quite a lot).

Some current CLI help functionality is broken compared to master, but presumably easy to fix (or potentially, we could change to a new way):
- `cylc --help` doesn't show the top level command help
- `cylc CATEGORY COMMAND --help`  shows category help instead of command help (I think we decided a long time ago though, that our command categories are not very useful).

Some commands (at least one) missing? 
- `cylc jobscript` - "unknown utility"

Sounds like you could use `click` more extensively? (The small bit in the main script looks nice). 

Seems like very minor problems.

@cylc/core - if others agree, shall we suggest @sgaist puts up a PR from his branch? 
- last commit "black"s all scripts - need to agree on that, but I imagine we would agree?
- [click](https://click.palletsprojects.com/en/7.x/) seems really nice, with good documentation; and it has a good [pedigree](https://palletsprojects.com/) - shall we just use it?  (Do we really need an exhaustive evaluation of all CLI arg parser tools - I vote not?)

**Comment by kinow** [comment]
Great work @sgaist !

On the package name, what about `cylc.flow.cli`, instead of `cylc.flow.scripts`? 

>last commit "black"s all scripts - need to agree on that, but I imagine we would agree?

I think it would be better to postpone it. We had discussed `black` before, and also the 100 characters limit. `black`'s default is 88, we are using 80, the pep had been updated to 100... I think in the end we disagreed on the length and decided to keep things as-is. Discussion on Riot

The `black` changes could cause issues if we have to revert changes. But I agree adopting it could potentially improve our code quality, reducing maintenance and helping contributors.

> click seems really nice, with good documentation; and it has a good pedigree - shall we just use it? (Do we really need an exhaustive evaluation of all CLI arg parser tools - I vote not?)

+1 from me.

**Comment by kinow** [comment]
Just tested the code, `cylc` itself has a different output, but my old workflow five worked like a charm with `cylc run --no-detach --verbose --debug five`. Amazing work @sgaist ! 

Soon I might be able to invoke some of the Cylc subcommands from Jupyter notebooks :)

**Comment by hjoliver** [comment]
> On the package name, what about cylc.flow.cli, instead of cylc.flow.scripts?

:+1:

**Comment by TomekTrzeciak** [comment]
> > click seems really nice, with good documentation; and it has a good pedigree - shall we just use it? (Do we really need an exhaustive evaluation of all CLI arg parser tools - I vote not?)
> 
> +1 from me.

Click is nice, but we have recently decided for [Clize](https://github.com/epsy/clize) in metoppv/improver#955 for our CLIs if you want to see how it looks like. The main attraction for us is integration with Sphinx (needed some hacking to support Google-style docstrings though).

**Comment by oliver-sanders** [comment]
Cool, please raise a PR when ready.

> last commit "black"s all scripts - need to agree on that, but I imagine we would agree?

Not keen on this personally but if you're all convinced, ignore me.
(if we really want to do this it would be best done in a quiet time when fewer conflicts would arise)

> click seems really nice, with good documentation; and it has a good pedigree - shall we just use it? (Do we really need an exhaustive evaluation of all CLI arg parser tools - I vote not?)

As far as I understand click is just syntatic sugar to optparse so fine. Just two requirements I'd like to flush out first:
* [ ] Need to be able to build a CLI programmatically (which I think we can since its really just optparse).
* [ ] Need Sphinx integration.

> On the package name, what about cylc.flow.cli, instead of cylc.flow.scripts?

:+1: and shift `cylc.flow.terminal` and `cylc.flow.option_parser` (if not removed) there too.

**Comment by oliver-sanders** [comment]
> The main attraction for us is integration with Sphinx (needed some hacking to support Google-style docstrings though).

I was about to ask just that, how did you get Google-style docstrings to work?

https://github.com/epsy/clize/issues/3

**Comment by TomekTrzeciak** [comment]
> > The main attraction for us is integration with Sphinx (needed some hacking to support Google-style docstrings though).
> 
> I was about to ask just that, how did you get Google-style docstrings to work?
> 
> [epsy/clize#3](https://github.com/epsy/clize/issues/3)

By subclassing ClizeHelp to insert docstring conversion in the right place:
https://github.com/neilCrosswaite/improver/blob/5645635/lib/improver/cli/__init__.py#L56-L95

**Comment by oliver-sanders** [comment]
> By subclassing ClizeHelp to insert docstring conversion in the right place

Nice, you should push that to epsy/clize.

**Comment by hjoliver** [comment]
Arg parser compartive notes: https://clize.readthedocs.io/en/stable/alternatives.html

After an admittedly pretty superficial look, I vote for click. The decorators are cool. It is also more active as a project. A quick google for sphinx integration found this: https://github.com/click-contrib/sphinx-click  (@oliver-sanders better to judge than me though, on this).

**Comment by matthewrmshin** [comment]
(So much for [The Zen of Python](https://www.python.org/dev/peps/pep-0020/).)

> There should be one-- and preferably only one --obvious way to do it.

**Comment by hjoliver** [comment]
I guess that only applies to the core language :rofl:

**Comment by hjoliver** [comment]
@sgaist - if you can reassure us on the two check-box concerns discussed above https://github.com/cylc/cylc-flow/issues/2802#issuecomment-541583721 then please go ahead and post a PR, but back out your commit of black-processed files (which is still not agreed for Cylc).  Also, if it is the case that click requires very minimal coding (as appears to be the case on your branch) we could easily swap it out again if necessary, before Cylc-8 is released (mid 2020), so that's not a big deal I think.

**Comment by sgaist** [comment]
About the missing commands, there currently are three:
- graph-diff
- jobscript
- scp-transfert

These are shell scripts that must be ported back to python but I wanted to have a proof of concept working before doing the port.

I haven't tested click-sphinx but it looks exactly what would be needed.

One thing I'd like to also note is that the current cylc script is just forwarding calls further mostly like the original shell script did. I have done it like that for the proof of concept.

If click is deemed the way to go, all the other commands could be ported to click as well and then plug in as sub-commands. That way help and parameters handling would be forwarded to the correct command. It will also allow for other cylc projects to add new command without having to modify the main package to support them.

For example, the command_list generated at beginning of the script can be avoided when all commands are implemented through click.

**Comment by hjoliver** [comment]
> These are shell scripts that must be ported back ... 

Ah, roger that.

> If click is deemed the way to go, ...

That would indeed be good (and what I was getting at with "Sounds like you could use click more extensively?" above) - can be done as a follow-up to the initial PR.

**Comment by oliver-sanders** [comment]
Closed by #3413
